internal/build/buildkit_printer.go

```diff
@@ -10,84 +10,112 @@ import (
 
 type buildkitPrinter struct {
 	output io.Writer
-	vMap   map[digest.Digest]*vertexAndLogs
+	vData  map[digest.Digest]*vertexAndLogs
+	vOrder []digest.Digest
 }
 
 type vertex struct {
-	digest digest.Digest
-	name   string
-	error  string
+	digest     digest.Digest
+	name       string
+	error      string
+	started    bool
+	completed  bool
+	cmdPrinted bool
 }
 
-type vertexLog struct {
-	vertex digest.Digest
-	msg    []byte
+// HACK: The prefix we assume here isn't valid for RUNS in exec format
+// Ex: RUN ["echo", "hello world"]
+const cmdPrefix = "/bin/sh -c "
+const buildPrefix = "    ╎ "
+
+func (v *vertex) isRun() bool {
+	return strings.HasPrefix(v.name, cmdPrefix)
+}
+
+func (v *vertex) isError() bool {
+	return len(v.error) > 0
 }
 
 type vertexAndLogs struct {
 	vertex *vertex
 	logs   []*vertexLog
 }
 
-var logDigest digest.Digest
+type vertexLog struct {
+	vertex digest.Digest
+	msg    []byte
+}
 
 func newBuildkitPrinter(output io.Writer) *buildkitPrinter {
 	return &buildkitPrinter{
 		output: output,
-		vMap:   map[digest.Digest]*vertexAndLogs{},
+		vData:  map[digest.Digest]*vertexAndLogs{},
+		vOrder: []digest.Digest{},
 	}
 }
 
-func (b *buildkitPrinter) parse(vertexes []*vertex, logs []*vertexLog) {
-
+func (b *buildkitPrinter) parseAndPrint(vertexes []*vertex, logs []*vertexLog) error {
```

Comment:
New strategy:

Collect all output for this "chunk". Then at the end figure out what, if anything, we should print given what we now know from the Docker daemon.